In [6]:
dependencies=[(2,1),(2,3)]

distances=[]
head_final_count=0

for head,dep in dependencies:
    dist=abs(head-dep)
    distances.append(dist)
    if head>dep:
        head_final_count+=1

MDD=sum(distances)/len(distances)
HF=head_final_count/len(dependencies)

print("依存距离列表:",distances)
print("平均依存距离(MDD):",MDD)
print("head-final比例(HF):",HF)

依存距离列表: [1, 1]
平均依存距离(MDD): 1.0
head-final比例(HF): 0.5


In [9]:
import spacy

# 加载英文模型
nlp = spacy.load("en_core_web_sm")

# 句子
doc = nlp("The cat sat.")

# 自动提取依存关系
dependencies = []
for token in doc:
    if token.dep_ != "ROOT":   # 忽略根节点
        head_pos = token.head.i + 1
        dep_pos = token.i + 1
        dependencies.append((head_pos, dep_pos))

print("自动提取的依存关系 (head, dependent):", dependencies)

# 计算 MDD 和 HF
distances = []
head_final_count = 0
for head, dep in dependencies:
    dist = abs(head - dep)
    distances.append(dist)
    if head > dep:
        head_final_count += 1

MDD = sum(distances) / len(distances)
HF = head_final_count / len(dependencies)

print("依存距离列表:", distances)
print("平均依存距离 (MDD):", MDD)
print("head-final 比例 (HF):", HF)

自动提取的依存关系 (head, dependent): [(2, 1), (3, 2), (3, 4)]
依存距离列表: [1, 1, 1]
平均依存距离 (MDD): 1.0
head-final 比例 (HF): 0.6666666666666666


In [10]:
import spacy
nlp = spacy.load("en_core_web_sm")

# 许渊冲译文
xu = "From hill to hill no bird in flight; From path to path no man in sight. A lonely fisherman afloat, Is fishing snow in lonely boat."

doc = nlp(xu)

# 自动提取依存关系
dependencies = []
for token in doc:
    if token.dep_ != "ROOT":
        head_pos = token.head.i + 1
        dep_pos = token.i + 1
        dependencies.append((head_pos, dep_pos))

# 计算 MDD 和 HF
distances = []
head_final_count = 0
for head, dep in dependencies:
    dist = abs(head - dep)
    distances.append(dist)
    if head > dep:
        head_final_count += 1

MDD = sum(distances) / len(distances)
HF = head_final_count / len(dependencies)

print(f"许渊冲译文：依存关系总数 = {len(dependencies)}")
print(f"平均依存距离 (MDD) = {MDD:.2f}")
print(f"head-final 比例 (HF) = {HF:.2f}")

许渊冲译文：依存关系总数 = 28
平均依存距离 (MDD) = 2.07
head-final 比例 (HF) = 0.50


In [13]:
import spacy
nlp_zh = spacy.load("zh_core_web_sm")

# 中文原文（柳宗元）
zh_text = "千山鸟飞绝，万径人踪灭。孤舟蓑笠翁，独钓寒江雪。"

doc_zh = nlp_zh(zh_text)

# 提取依存关系
deps_zh = []
for token in doc_zh:
    if token.dep_ != "ROOT":
        head_pos = token.head.i + 1
        dep_pos = token.i + 1
        deps_zh.append((head_pos, dep_pos))

# 计算 MDD 和 HF
distances_zh = []
hf_count_zh = 0
for h, d in deps_zh:
    dist = abs(h - d)
    distances_zh.append(dist)
    if h > d:
        hf_count_zh += 1

MDD_zh = sum(distances_zh) / len(distances_zh)
HF_zh = hf_count_zh / len(deps_zh)

print(f"中文原文：依存关系数={len(deps_zh)}, MDD={MDD_zh:.2f}, HF={HF_zh:.2f}")

中文原文：依存关系数=12, MDD=2.50, HF=0.33


In [22]:
import spacy
nlp_en = spacy.load("en_core_web_sm")

snyder = "These thousand peaks cut off the flight of birds, On all the trails, human tracks are gone. A single boat — coat — hat — an old man! Alone fishing chill river snow."

doc = nlp_en(snyder)

# 提取依存关系
deps = []
for token in doc:
    if token.dep_ != "ROOT":
        head_pos = token.head.i + 1
        dep_pos = token.i + 1
        deps.append((head_pos, dep_pos))

# 计算 MDD 和 HF
distances = []
hf_count = 0
for h, d in deps:
    dist = abs(h - d)
    distances.append(dist)
    if h > d:
        hf_count += 1

MDD = sum(distances) / len(distances)
HF = hf_count / len(deps)

print(f"斯奈德译文：依存关系数 = {len(deps)}")
print(f"平均依存距离 (MDD) = {MDD:.2f}")
print(f"head-final 比例 (HF) = {HF:.2f}")

斯奈德译文：依存关系数 = 35
平均依存距离 (MDD) = 2.51
head-final 比例 (HF) = 0.57


# 依存距离分析报告：以《江雪》两个英译本为例

## 1. 目的
用论文中的 **平均依存距离 (MDD)** 和 **head-final 比例 (HF)** 两个指标，量化对比《江雪》中文原文及两个英译本的句法结构差异。

## 2. 方法
- 用 `spaCy` 提取每个句子的依存关系。
- 计算 **MDD**（所有依存距离的平均值，数值越大句子结构越复杂）。
- 计算 **HF**（支配词出现在从属词之后的比例，HF>0.5 倾向于“头在后”语序）。

## 3. 数据结果

| 文本 | MDD | HF |
|------|-----|-----|
| 中文原文 | 2.50 | 0.33 |
| 许渊冲译文 | 2.07 | 0.50 |
| 斯奈德译文 | 2.51 | 0.57 |

## 4. 分析结论
- **许渊冲译文**：MDD 较低（2.07），句子结构更简单、更符合英语习惯；HF 正好 0.50，属于典型的英语均衡型。说明他倾向于“归化”策略。
- **斯奈德译文**：MDD（2.51）几乎与中文原文（2.50）相同，保留了原诗的句法复杂度；HF 为 0.57，略偏向“头在后”，接近中文的语序倾向。说明他更注重“异化”和忠实于意象。

## 5. 下一步
- 扩大语料（更多诗句或散文）验证结论。
- 尝试计算依存方向的其他指标（如 head-initial 比例）。
- 与论文中的多语言对比结果进行参照。